### DISCUSSION TITLE
2FA with an enforced factor and an optional one

### DISCUSSION THREAD

**Discussion by VladNastase** [keycloak/keycloak#12678] [Q&A]
I want to have a login flow that enforces 2FA via OTP, but allows the users to configure and use other second factor options, such as Webauthn if they desire.
Currently, the flow looks like this:
![1](https://user-images.githubusercontent.com/22620488/175271043-a2582465-8db8-4c41-97be-2c165f75e3f7.png)
which works fine (every time a user with an unconfigured OTP 2FA logs in, it will redirect the user to the setup page).
### Option 1
![2](https://user-images.githubusercontent.com/22620488/175271148-094c2ee2-d494-4dcc-a0fe-31030010518b.png)

Any attempt to login without either OTP or Webauthn configured fails with the error "Invalid credentials".
This happens even if I manually assign the user the "Configure OTP" or "Webauthn register" required actions.

I've traced this to the following lines:
https://github.com/keycloak/keycloak/blob/92c4e6d585dca4d2f741ac4ac6b50ea0c0940a9a/services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java#L276-L281
https://github.com/keycloak/keycloak/blob/92c4e6d585dca4d2f741ac4ac6b50ea0c0940a9a/services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java#L439-L449

It looks like required actions attached to an execution are not redirected to/executed if the execution is not required (in this case it's alternative).
Did I understand this right and is there a reason for this behavior?

### Option 2
![3](https://user-images.githubusercontent.com/22620488/175271211-b328a1ea-04b1-4630-8d0c-6321aecd5723.png)

This enforces OTP, but doesn't offer Webauthn as an alternative at login (if configured). I presume that for this to happen, executions need to be inside the same flow/subflow, but I didn't look into it.

Is there any way to construct a flow that can achieve enforcing one type of 2FA while allowing the user to configure a second type?

**Comment by jackie-linz** [comment]
This should meet your need

```yaml
- Username password form (REQUIRED)
- 2FA (Flow - REQUIRED)
  - Authenticate with 2FA if configured (Flow - ALTERNATIVE)
    - Condition - user configured (REQUIRED)
    - OTP form (ALTERNATIVE)
    - WebAuthen (ALTERNATIVE)
  - Register OTP (Flow - ALTERNATIVE)
    - OTP form (REQUIRED)
```

see discussion: https://github.com/keycloak/keycloak/discussions/14988
